File: Q-2050 - Parallel Courses III/2050.py
class Solution:
    def minimumTime(self, n: int, relations: List[List[int]], time: List[int]) -> int:
        
        graph = defaultdict(set)
        degree = defaultdict(int)
        for i, j in relations:
            degree[j-1] += 1
            graph[i-1].add(j-1)
            
        count = 0        
        prev = defaultdict(int)
        
        queue = [(time[x], x) for x in range(n) if degree[x] == 0]
        heapify(queue)
        
        while(queue):
            temp, course = heappop(queue)
            count += 1
            if(count == n):
                return temp
            for nei in graph[course]:
                degree[nei] -= 1
                prev[nei] = max(prev[nei], temp)
                if(degree[nei] == 0):
                    heappush(queue, (prev[nei] + time[nei], nei))

File: Q-2050 - Parallel Courses III/2050.cpp
class Solution {
public:
    int minimumTime(int n, vector<vector<int>>& relations, vector<int>& time) {
        
        vector<vector<int>> graph(n + 1);
        for(auto i : relations)
            graph[i[1]].push_back(i[0]);

        vector<bool> done(n + 1, false);
        int answer = 0;
        for(int i = 1; i <= n; i++)
            if(!done[i])
                search(i, graph, time, done, answer);

        return answer;
    }

    int search(int i, vector<vector<int>> &graph, vector<int> &time, vector<bool> &done, int &answer)
    {
        if(done[i])
            return time[i - 1];

        done[i] = true;
        int temp = 0;
        for(auto j : graph[i])
            temp = max(temp, search(j, graph, time, done, answer));

        time[i - 1] += temp;
        answer = max(answer, time[i - 1]);
        return time[i - 1];
    }
};